### PR TITLE
add checksum to ChangeLog

### PR DESCRIPTION
**Pull request by YLiLarry**

### CODE REVIEW

File: src/frontend/src/helpers/Storage/LocalStorage.ts [hunk 1]
```diff
@@ -424,7 +429,7 @@ class LocalStorage implements AbstractStorage {
         if (change.type === "deleteFile") {
           await this.deleteFile(fid, false);
         } else if (change.type === "editFile") {
-          await this.writeFile(fid, change.contents, false);
+          await this.writeFile(fid, change.contents, undefined, false);
```
Comment (e45lee) [inline src/frontend/src/helpers/Storage/LocalStorage.ts]: you want change.checksum here (online checksum from the server)

**Comment by YLiLarry** [comment]
All tests fixed and passed. Could you check again and merge it?
Thanks.

File: src/frontend/src/helpers/Storage/WebStorage.ts [hunk 2]
```diff
@@ -117,16 +114,22 @@ class WebStorage extends AbstractStorage implements AbstractWebStorage {
   };
 
   public async newProject(name: string): Promise<ProjectBrief> {
-    if (this.offlineMode === OfflineMode.On ||
-        this.offlineMode === OfflineMode.Forced) {
-      return this.storage.newProject(name);
-    }
+    // if (this.offlineMode === OfflineMode.On ||
+        // this.offlineMode === OfflineMode.Forced) {
+    const off = this.storage.newProject(name);
```
Comment (e45lee) [inline src/frontend/src/helpers/Storage/WebStorage.ts]: I'd do a sync instead afterwards (as you need to fetch the template files)

File: src/frontend/src/helpers/Storage/WebStorage.ts [hunk 3]
```diff
@@ -417,7 +420,7 @@ class WebStorage extends AbstractStorage implements AbstractWebStorage {
     }
     const changesSent = await this.storage.getChangeLogs();
     const settingsSent = await this.storage.getSettings();
-    if (!fetchServerChanges) {
+    if (! fetchServerChanges) {
       if (changesSent.length === 0) {
```
Comment (e45lee) [inline src/frontend/src/helpers/Storage/WebStorage.ts]: We should also test if we've made any changes to project settings, but that can wait.  The easiest way to do this is to have a localStorage variable that's true if and only if we've made any changes.

**Pull request merged**